### PR TITLE
Clamp the intermediate value in bce/backward lowering to avoid nan

### PR DESCRIPTION
**Pull request by JackCaoG**
This is to fix https://github.com/pytorch/xla/issues/2446. Pytorch clamp its value during forward in https://github.com/pytorch/pytorch/blob/ba98c0e38ca76a8e03d27e20a1d2bfc324d502ca/aten/src/ATen/native/Loss.cpp#L147 and clamp its value during backward in https://github.com/pytorch/pytorch/blob/ba98c0e38ca76a8e03d27e20a1d2bfc324d502ca/aten/src/ATen/native/Loss.cpp#L185

### CODE REVIEW

**Comment by taylanbil** [comment]
Thanks Jack!

File: test/test_operations.py [hunk 1]
```diff
@@ -664,6 +664,21 @@ def test_get_xla_tensor(self):
     self.assertEqual(tx, sx.data.cpu())
 
 
+class TestBinaryCrossEntropyLimitValue(XlaTestCase):
+
+  def test_cross_entropy_loss(self):
+    lossfn = nn.BCELoss()
+    target_cpu = torch.tensor((1.))
```
Comment (davidel) [inline test/test_operations.py:671]: We have an existing harness to test XLA vs CPU:

https://github.com/pytorch/xla/blob/b1e4f93652516e9c497b44e76adf69fe1bc860eb/test/test_operations.py#L1588

File: torch_xla/csrc/reduction.cpp [hunk 2]
```diff
@@ -137,8 +137,11 @@ xla::XlaOp BuildBinaryCrossEntropy(xla::XlaOp input, xla::XlaOp target,
         XlaHelpers::ScalarBroadcast<float>(1.0, input_shape, target.builder());
   }
   xla::XlaOp one = xla::One(input.builder(), input_shape.element_type());
-  xla::XlaOp result = -xweight * (target * xla::Log(input) +
-                                  (one - target) * xla::Log(one - input));
+  xla::XlaOp log_bound = XlaHelpers::ScalarValue(
+      -100, input_shape.element_type(), input.builder());
```
Comment (davidel) [inline torch_xla/csrc/reduction.cpp:141]:
```c++
static const float kLogBound = -100;
```

File: torch_xla/csrc/reduction.cpp [hunk 3]
```diff
@@ -164,7 +167,10 @@ xla::XlaOp BuildBinaryCrossEntropyBackward(
         XlaHelpers::ScalarBroadcast<float>(1.0, input_shape, target.builder());
   }
   xla::XlaOp one = xla::One(input.builder(), input_shape.element_type());
-  xla::XlaOp result = xweight * (input - target) / input / (one - input);
+  xla::XlaOp epsilon = XlaHelpers::ScalarValue(
+      1e-12, input_shape.element_type(), input.builder());
```
Comment (davidel) [inline torch_xla/csrc/reduction.cpp:171]:
```c++
static const float kEpsilon = 1e-12;
```

**Review by davidel — CHANGES_REQUESTED** [review]

File: torch_xla/csrc/reduction.cpp (hunk 2 above)
Comment (JackCaoG) [reply torch_xla/csrc/reduction.cpp]: out of curiosity, when do we prefix a variable with `k`? I have seen this many times in our code base but didn't know why.

File: test/test_operations.py [hunk 4]
```diff
@@ -664,6 +670,20 @@ def test_get_xla_tensor(self):
...
+
+  def test_cross_entropy_loss(self):
+
+    def test_fn(pred, target, offset):
+      lossfn = nn.BCELoss()
+      return lossfn(pred - offset, target)
+
+    pred = torch.tensor(1.0)
+    target = torch.tensor(1.0)
+    for offset in [1, 0, 1e-8, 1e-7]:
```
Comment (davidel) [inline test/test_operations.py:683]: There is no need for the kwargs change above.
Just move the function in the loop and let it capture the current context.
